extras: dynamic bash completion

fredericmoulins opened the pull request:
Move bash-completion script for the `gluster` command
in `/usr/share/bash-completion/completions`
instead of `/etc/bash_completion.d`.

Allow bash-completion to dynamically load the script
only when a user tries to tab complete on the `gluster` command.

Fixes: #3750
Change-Id: Ifca300bdf5b14a68726a6fdfdcc19052cf624f85
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

fredericmoulins committed:
extras: dynamic bash completion

Move bash-completion script for the `gluster` command
in `/usr/share/bash-completion/completions`
instead of `/etc/bash_completion.d`.

Allow bash-completion to dynamically load the script
only when a user tries to tab complete on the `gluster` command.

Fixes: #3750
Change-Id: Ifca300bdf5b14a68726a6fdfdcc19052cf624f85
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

mykaul commented:
/run regression

amarts reviewed (dismissed):
LGTM

Would be great someone with more knowledge on rpm packages looks at it once.

xhernandez commented:
There's a question in bash completion faqs (https://github.com/scop/bash-completion/blob/master/README.md#faq) regarding how to maintain completion code for an external package.

It provides a more dynamic way of detecting the right location for the file depending on the build system. I think we should use the recommended method.

fredericmoulins committed:
extras: generate command-completion Makefile

Use pkg-config file variable as recommended by bash-completion [1]
to generate command-completion Makefile and README for source distribution.

The `Makefile.in` and `README.in` are explicitely included in the source
distribution. `command-completion` is still not considered a subdir of
`extras` while building so that the bash completion script is not
installed.

[1]: https://github.com/scop/bash-completion#faq

Change-Id: Ib06c638f54f1eccfb1b1ce2e5aab8919b9b2e4f0
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

fredericmoulins committed:
spec: use pkgconfig bash-completion file variable

Use pkgconfig to get the bash-completion scripts directory. [1]

[1]: https://github.com/scop/bash-completion#faq

Change-Id: Ibc8fc96e0acf966fe7849f7a1d995982ac79d62e
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

fredericmoulins committed:
spec: set bash-completion script as non executable

Change-Id: I7c219b6aff034ba99ce0a6ca2f486529fbb15e05
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

fredericmoulins committed:
spec: add BuildRequire bash-completion for cli

The cli bash-completion script installation tries to get the
installation directory for completion scripts from pkgconfig.

Change-Id: Icb423299e913e2f534068011b2adc3d3965921ed
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

fredericmoulins commented:
Hi,

thank you for your reviews.

> It provides a more dynamic way of detecting the right location for the file depending on the build system. I think we should use the recommended method.

I added commits to the PR to try to do that in the `configure.ac` and in the spec file. The `PKG_CHECK_VAR` macro doesn't seem to be available in the smoke-centos7 and devrpm-el7 builds.

From the comments in the `aclocal.m4` generated on my machine, the `PKG_CHECK_VAR` macro seems to exist `Since: 0.28`. I suppose it is the version of [pkg-config](https://repology.org/project/pkg-config/versions) (not `pkgconf`) which is at version 0.27.1 in CentOS7, and doesn't seem in the EPEL7 repos. On Fedora, the macro is provided by the [pkgconf](https://repology.org/project/pkgconf/versions) package (`pkgconf-m4`) which has the macro since 0.8.10 (all packaged versions seem to have it).

Do you have any hint on how to handle this or how to guard the macro?

xhernandez commented on configure.ac:
```diff
@@ -1609,6 +1611,10 @@ if test x$BUILD_READLINE = xyes; then
     AC_DEFINE(HAVE_READLINE, 1, [Enable CLI readline support.])
 fi
 
+PKG_CHECK_VAR(bashcompdir, [bash-completion], [completionsdir], ,
+  bashcompdir="${sysconfdir}/bash_completion.d")
+AC_SUBST(bashcompdir)
```
I think this should work as an alternative for `PKG_CHECK_VAR`. You can also add a comment saying that `PKG_CHECK_VAR` should be used but it's not available in some distributions.

```suggestion
bashcompdir="$(pkg-config --variable=completionsdir bash-completion)"
if test "x${bashcompdir}" = "x"; then
    bashcompdir="${sysconfdir}/bash_completion.d")
fi
```

fredericmoulins replied:
Yes, thanks. This seems simpler than more macros. I will update based on that.

xhernandez commented on glusterfs.spec.in:
```diff
@@ -334,6 +338,7 @@ and client framework.
 
 %package cli
 Summary:          GlusterFS CLI
+BuildRequires:    pkgconfig(bash-completion)
```
I'm not sure if we really need to force this dependency.

fredericmoulins replied:
I am not sure either.

For systems where bash-completion and its pkgconf are easily available, I guess it should: to be able to install the script the recommended way so that it is loaded dynamically. For systems where it might not be the case, the default in `/etc/bash-completion.d` is fine anyway, so the requirement is not needed.

Could a condition be added around the BuildRequires?

xhernandez replied:
I don't know if it can be done conditionally.

The problem I see is that if the requirement cannot be satisfied for some reason, the build will fail, when it shouldn't. Not sure what's the right way to manage this.

fredericmoulins replied:
I agree that the package build should not break on that. If it is not available the default compatible path is enough.

The pkg-config definitions were added in bash-completion in version 1.90 [[1]]. Supported distros seem to have 2.x versions. (The devrpms-el7 job build it ok with the new path.)

[1]: https://github.com/scop/bash-completion/commit/90b67bc976dccc655f7ba89714b6c1e5b1e5ddc5

There are spec macros to check on distribution and distribution version. There are some used in the spec file (eg for systemd).
Maybe a condition for RHEL would be needed, but I don't know how to check that.

nixpanic replied:
As this is 'only' a `BuildRequires`, it will only be needed to be installed while building RPMs. If for some reason it is not available on a distribution, it should be simple for a packager to provide a workaround for that particular distribution.

Shwetha-Acharya replied:
If we determine distros or any other parameters where we have the required package available for sure, then we can add a conditional similar to: https://github.com/gluster/glusterfs/blob/devel/glusterfs.spec.in#L291

Also adding a comment explaining that the default in /etc/bash-completion.d would work incase the build requires of this package cannot be met due to its unavailability in any system would help a packager to modify this requirement

fredericmoulins replied:
Thank you for your review.

I added a comment and a condition to exclude old RHEL versions as an example. Most rpm distributions that are supported today have recent enough version for bash-completion, so the idea is to exclude some old distribution versions instead of specifically allowing most of the current ones. I hope it is ok.

xhernandez commented on glusterfs.spec.in:
```diff
@@ -224,6 +224,10 @@
 %global __provides_exclude_from ^%{_libdir}/glusterfs/%{version}/.*$
 %endif
 
+%global bashcompdir %(pkg-config --variable=completionsdir bash-completion 2>/dev/null)
+%if "%{bashcompdir}" == ""
+%global bashcompdir ${sysconfdir}/bash_completion.d
+%endif
```
Isn't this what configure.ac is already doing ?

fredericmoulins replied:
Yes, it is pretty much the same thing.

The first thing I did was to use the `@bashcompdir@` configure macro. The generated `glusterfs.spec` then contained the full path kind of hardcoded. It looked weird to me. I guess it is the idea that if someone takes the `glusterfs.spec` to build an rpm, the path should be determined dynamically at the time, and on the system, of the package build, versus at the time the source distribution was done. Does that make sense?

I saw that very recently shell completion dirs rpm macros were added in Fedora [[1]]. I guess it is the same idea but here with a dynamic pkg-config call.

[1]: https://src.fedoraproject.org/rpms/redhat-rpm-config/c/7663a47b1a48e55c17c92d08af937e340c66df86?branch=f36

xhernandez replied:
But the `glusterfs.spec` can only be generated by `configure` from `glusterfs.spec.in`, so it's already tailored for a specific environment.

Anyway I'm not an expert here and I don't know what's the best option.

fredericmoulins replied:
I also thought of that because the final `glusterfs.spec` (as well as the `glusterfs.spec.in) is part of the source distribution archive.

nixpanic replied:
Doing the detection (again) in the generated `glusterfs.spec` is a clean solution IMHO. In case a distribution changes the location of the bash-completion (or different RPM distributions have different locations), the same dist tarball does not need to be adapted.

xhernandez commented on extras/Makefile.am:
```diff
@@ -39,15 +39,17 @@ scripts_SCRIPTS += control-mem.sh
 endif
 endif
 
+CLEANFILES = command-completion/Makefile command-completion/README
```
Not sure if this is necessary

fredericmoulins replied:
As I understand it, the Makefile is not run during install, the files are not taken into account for a dist clean. Please see my comment on the Readme.in file.

xhernandez replied:
Understood.

xhernandez commented on extras/Makefile.am:
```diff
@@ -39,15 +39,17 @@ scripts_SCRIPTS += control-mem.sh
-	pre-upgrade-script-for-quota.sh command-completion/gluster.bash \
-	command-completion/Makefile command-completion/README \
+	pre-upgrade-script-for-quota.sh \
+	command-completion/gluster.bash command-completion/README.in \
+	command-completion/Makefile.am command-completion/Makefile.in \
```
Not sure if this is necessary

fredericmoulins replied:
As I understand it, the Makefile is not run during install, the files are not taken into account as dist files. Please see my comment on the Readme.in file.

xhernandez replied:
Understood.

nixpanic replied:
did you consider adding `command-completion` to `SUBDIRS`? That might be cleaner and gives you the option to do include a `install-data-local` target so that `make install` does it all automatically, no need to manually copy things around in the spec file.

fredericmoulins replied:
I just pushed a commit to include `command-completion` in the `SUBDIRS` for extras, so that the bash completion is installed by default. It is simpler overall, indeed.

During an installation from source, the cli seems to always be installed, so I guess the bash-completion script can be installed by default too. It was not the case before this PR and this commit: you can include the commit if you think it is ok.

xhernandez commented on extras/command-completion/README.in:
```diff
@@ -0,0 +1,7 @@
+This file is not moved to @bashcompdir@
+with the source installation.
+Please execute make install explicity from here to move this file to
+@bashcompdir@
+
+Similarly, use make uninstall to remove it from
+@bashcompdir@
```
I think this is not true anymore since we have added the Makefile into configure.ac, so it will be installed and uninstalled as all other components. Maybe we should just remove this file.

fredericmoulins replied:
The way I understand it is that the `Makefile` and `README` files are generated by the configure step from the declarations in configure.ac, but the `Makefile` is not included from the `extras/Makefile` during the make step (`command-completion` not declared as a subdir), so the files are not included in dist files. Unless I am missing something in configure, I think that is also why I had to add the '.in'  and '.am' files explicitly to the dist files in `extras/Makefile`.

So as it is, the `extras/command-completion/Makefile` is still not run during make install. It could probably be. I tried to keep the same behaviour, as I don't know why the completion script should or should not be installed for a source installation. The Makefile is now an automake generated one instead of a short simple one, but it does basically the same thing.

xhernandez replied:
I think you are right. I assumed adding them to `configure.ac` was enough, but I forgot that the directory also needs to be added to some `Makefile.am` to be automatically processed.

xhernandez commented:
@kalebskeithley could you take a look, please ? I'm not as familiarized as you with spec files and its best practices.

fredericmoulins committed:
extras: get bashcompdir macro from pkg-config

PKG_CHECK_VAR is not available on EL7 and CentOS7.

Change-Id: I1ea864b4540566b7b3072f2a049746097eddfb3a
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

nixpanic commented on glusterfs.spec.in:
```diff
@@ -980,8 +985,8 @@ touch %{buildroot}%{_sharedstatedir}/glusterd/nfs/run/nfs.pid
 ## Install bash completion for cli
-install -p -m 0744 -D extras/command-completion/gluster.bash \
-    %{buildroot}%{_sysconfdir}/bash_completion.d/gluster
+install -p -m 0644 -D extras/command-completion/gluster.bash \
```
you might want to do this in the `extras/command-completion/Makefile.in` (but only if the CLI is enabled)?

fredericmoulins replied:
The CLI seems to always be installed and the completion script is installed as data (with `0644`) from the `extras/command-completion/Makefile`.

With the latest commit and the completion script installed by default, this install call is no longer needed in the spec file.

nixpanic reviewed (commented):
this looks reasonable to me. It might be possible to simplify things a little more, but you would need to test it out. If this works fine, I'd approve it as it can always get improved when needed later on.

fredericmoulins committed:
extras: install command completion by default

Add `command-completion` as subdir in `extras/Makefile.am`.

Remove the specific make and spec rules for distribution and
installation of the command-completion files.

Remove the `README` with the notice for manual installation.

Change-Id: I6e8bf423ea1f51bb7e7358da270dbb0e75d79e48
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

xhernandez commented:
/run regression

xhernandez reviewed (commented):
LGTM. Just address Shwetha's concern.

gluster-ant commented:
1 test(s) failed 
./tests/basic/fuse/active-io-graph-switch.t

0 test(s) generated core 


6 test(s) needed retry 
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
./tests/basic/fuse/active-io-graph-switch.t

5 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
https://build.gluster.org/job/gh_centos7-regression/2931/

fredericmoulins committed:
spec: guard and comment cli BuildRequire

Change-Id: I726e840e701f3e7f21971a947636b346566fff25
Updates: #3750
Signed-off-by: Frédéric Moulins <frederic@moulins.org>

Shwetha-Acharya commented:
/run regression
